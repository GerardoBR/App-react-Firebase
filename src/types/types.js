export const types={
    login : '[auth] Login',
    logout : '[auth] Logout',


    uiSetError : '[UI] set Error',
    uiRemoveError : '[UI] Remove Error',


    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',

    notesAddNew : '[Notes] New notes',
    notesActive : '[Notes] Set active notes',
    notesLoad : '[Notes] Load notes',
    notesUpdated : '[Notes] Update notes saved',
    notesFileUrl : '[Notes] Update  image URL',
    notesDelete  : '[Notes] Delete  image URL',
    notesLogoutCleaning : '[Notes] Logout  Cleaning',


}