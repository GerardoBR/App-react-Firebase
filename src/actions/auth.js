import Swal from 'sweetalert2'
import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from "@firebase/auth";
import { googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading }  from '../actions/ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword=(email,password)=>{
    return(dispatch)=>{
        dispatch(startLoading());
        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
          .then((userCredential) => {
            // Signed in 
            const user = userCredential.user;
            // console.log(user)
            dispatch(login(user.uid, user.displayName));    
            dispatch(finishLoading());
            // ...
          })
          .catch((error) => {
            //   console.log( error);
              dispatch(finishLoading());
              Swal.fire('Error',error.message,'error');
            // const errorCode = error.code;
            // const errorMessage = error.message;
            // ..
          });
    }
}

export const startRegisterWithEmailPasswordName=(name, email,password)=>{
    return(dispatch)=>{

        const auth = getAuth();
            createUserWithEmailAndPassword(auth, email, password)
                .then( async({ user }) => {
 
                await updateProfile( user, { displayName: name });
                dispatch(login(user, user.displayName))    
            // console.log(user.uid, name);
        }).catch(e=>{
            let mns  = "";
            // console.log(e.message === 'Firebase: Error (auth/email-already-in-use).');
            if(e.message === 'Firebase: Error (auth/email-already-in-use).'){
                
                mns = "email-already-in-use";
            }
            Swal.fire('Error',mns,'error');
        });
    
    }
}
export const startGoogleLogin=()=>{

    return (dispatch)=>{
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) =>{
                // console.log(user);
                dispatch(login(user.uid, user.displayName))
            });
    }
}


export const login=(uid,displayName)=>({

        type :types.login,
        payload :{
            uid,
            displayName
        }
});

export const  Startlogout=()=>{
    return async ( dispatch )=>{
        const auth = getAuth();
           signOut(auth)
           dispatch(logout());
           dispatch(noteLogout());
        //    .then(() => {
        //         dispatch(logout());
        //         dispatch(noteLogout());
        //     }).catch((error) => {

        //         console.log(error);
        //     });
    }
}
export const logout =()=>({
    type : types.logout
})