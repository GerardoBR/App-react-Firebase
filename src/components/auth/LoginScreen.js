import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {  startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { remuveError, setError } from '../../actions/ui'
import  validator from 'validator';
export const LoginScreen = () => {

    const dispatch = useDispatch();

    const {loading} = useSelector( state=> state.ui );
    // console.log(loading);
    const [ formValue, handleInputChange ]=useForm({
        email:'gera2@gmail.com',
        password : '12345678'

    })
    const { email , password } =formValue;

    const handleLogin=(e)=>{
        // console.log("hey");
        e.preventDefault();
        // console.log(email,password)
        if(isFormValid()){
            // console.log("entro");
            dispatch(startLoginEmailPassword( email,password ));
           
        }
    }
    const handleGoogleLogin=()=>{
     
        dispatch(startGoogleLogin());
    }
    const isFormValid = ()=>{
        if(!validator.isEmail(email) ){
            console.log("email is not valid");
            dispatch(setError( "email is not valid" ));
            return false;
        }
        dispatch(remuveError());
        return true;
    }
    // const handleLoginPassword=(e)=>{
    //     e.preventDefault();
    //     dispatch(startLoginEmailPassword( email,password ));
    // }

    return (
        <div  className = "animate__animated animate__fadeIn animate__faster">
            <h3 className ="auth__title">Login</h3>
            <form 
               
            onSubmit ={ handleLogin }>
                <input
                    type = "text"
                    placeholder = "email"
                    name = "email"
                    className="auth__input"
                    autoComplete = "off"
                    value = { email }
                    onChange ={ handleInputChange }
                />
                <input
                    type = "password"
                    placeholder = "password"
                    name = "password"
                    className="auth__input"
                    value= { password }
                    onChange ={ handleInputChange }
                />
                <button 
                    type ="submit" 
                    className= "btn btn-primary btn-block"
                    disabled = { loading }
                      
                >
                Login
                </button>

                <div className ="auth__social-networks">
                    <p>Login Social Networks</p>
                    <div 
                        className="google-btn"
                        onClick = { handleGoogleLogin }
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link  
                    to ="/auth/register"
                    className =" link "
                >
                    Create new account
                </Link>

             

            </form>
        </div>
    )
}
