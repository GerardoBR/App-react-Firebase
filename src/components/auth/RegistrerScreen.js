import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux';
import  validator from 'validator';
import { remuveError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegistrerScreen = () => {

    const dispatch = useDispatch();
    const { msgError } = useSelector( state=> state.ui );
    // console.log(msgError);
    const [formValues, handleInputChange]= useForm({
        name : 'Gerardo2',
        email : 'gera2@gmail.com',
        password1 : 12345678,
        password2 : 12345678
    });

    const { name,email, password1,password2} = formValues;
    const handleRegister=(e)=>{
        // console.log(name,email,password1,password2);
        e.preventDefault();
        if(isFormValid()){
            // console.log('formulario correcto');
          dispatch(startRegisterWithEmailPasswordName(name,email, password1));
        }

    }
    const isFormValid = ()=>{
        if( name.trim().length === 0 ){
            // console.log("Name is required");
            dispatch(setError( "Name is required" ));
            return false;

        }else if(!validator.isEmail(email) ){
            // console.log("email is not valid");
            dispatch(setError( "email is not valid" ));

            return false;
        }else if ( password1 !== password2 || password1.length<8){
            // console.log("Password should be at last 6 characters")
            dispatch(setError( "Password should be at last 6 characters" ));

            return false;
        }
        dispatch(remuveError());
        return true;
    }

    return (
        <div className="animate__animated animate__fadeIn animate__faster">
            <h3 className ="auth__title">Register</h3>

            <form onSubmit = { handleRegister }>
                {
                    msgError &&
                    (
                        <div className = "auth__alert">
                            { msgError }
                        </div>
                    )
                }
            <input
                    type = "text"
                    placeholder = "Name"
                    name = "name"
                    className="auth__input"
                    autoComplete = "off"
                    value =  { name }
                    onChange ={ handleInputChange }
                />
                <input
                    type = "text"
                    placeholder = "email"
                    name = "email"
                    className="auth__input"
                    autoComplete = "off"
                    onChange ={ handleInputChange }
                    value =  { email }
                />
                <input
                    type = "password"
                    placeholder = "password"
                    name = "password1"
                    className="auth__input"
                    onChange ={ handleInputChange }
                    value =  { password1 }
                />
                 <input
                    type = "password"
                    placeholder = "Confirm password"
                    name = "password2"
                    className="auth__input"
                    onChange ={ handleInputChange }
                    value =  { password2 }
                />
                <button 
                    type ="submit" 
                    className= "btn btn-primary btn-block mb-5"
                    // disabled= { true }
                >
                Register
                </button>

                
                <Link  
                    to ="/auth/login"
                    className =" link "
                >
                    Already register?
                </Link>

             

            </form>
        </div>
    )
}
