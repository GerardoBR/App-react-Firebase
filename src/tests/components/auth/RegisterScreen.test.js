import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store'; //ES6 modules
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { RegistrerScreen } from "../../../components/auth/RegistrerScreen";
import { MemoryRouter } from 'react-router';
import { types } from '../../../types/types';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const intstate  =  {
    auth :{},
    ui : {
        loading:false,
        msgError:null
    },
    notes : {
        notes :[],
        active : null
    }

};
let  store = mockStore(intstate);

// store.dispatch = jest.fn();
const wrapper = mount(
<Provider store = { store }>
    <MemoryRouter>
        <RegistrerScreen/>
    </MemoryRouter>
</Provider> 
);

describe('Pruebas en el componente <RegisterScreen/> ', () => {


    // beforeEach(()=>{
    //     store = mockStore (intstate);
    //     // jest.clearAllMocks();
    // });

    test('debe mostrarse correctamente con el Snapshoot', () => {

        expect(wrapper).toMatchSnapshot();
    });
    test('debe de hacer el dispatch de la accion respectiva', () => {

        const emailFile = wrapper.find('input[name="email"]');
        
        emailFile.simulate('change',{
            target : {
                value : '',
                name : 'email'
            }
        })
        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });
        const action  = store.getActions();
        // console.log(action);
        expect( action[0]).toEqual({
            type: types.uiSetError,
            payload :'email is not valid'
        })

    });
    test('Debe de mostrar la alerta del error ', () => {
        
        const intstate  =  {
            auth :{},
            ui : {
                loading:false,
                msgError:'email is not valid'
            },
            notes : {
                notes :[],
                active : null
            }
        
        };
        const  store = mockStore(intstate);

        // store.dispatch = jest.fn();
        const wrapper = mount(
        <Provider store = { store }>
            <MemoryRouter>
                <RegistrerScreen/>
            </MemoryRouter>
        </Provider> 
        );
        

        expect(wrapper.find('.auth__alert').exists()).toBe(true);
        expect(wrapper.find('.auth__alert').text().trim()).toBe(intstate.ui.msgError);

    })
    
    

    
    
})
