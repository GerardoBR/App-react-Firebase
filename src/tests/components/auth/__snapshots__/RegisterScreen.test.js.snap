// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <RegisterScreen/>  debe mostrarse correctamente con el Snapshoot 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Gerardo2"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value="gera2@gmail.com"
    />
    <input
      className="auth__input"
      name="password1"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value={12345678}
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value={12345678}
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className=" link "
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>
</div>
`;
